Do not N^2 loading webhook configurations

lavalamp opened the pull request:
This adds a processing delay to deduplicate notifications to reload webhook configurations. Solves N^2 behavior on startup and prevents useless work when webhooks change in rapid succession.

#### What type of PR is this?

/kind bug
/kind cleanup

#### What this PR does / why we need it:

Fixes N^2 behavior I noticed while working on #113985.


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-apiserver: removed N^2 behavior loading webhook configurations.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

lavalamp commented:
I'll add corresponding changes to the mutating version of this file once someone looks at this.

lavalamp commented:
/sig api-machinery

lavalamp commented:
The test fail is real, the goroutine doesn't get cleaned up.

apelisse commented on staging/src/k8s.io/apiserver/pkg/admission/configuration/validating_webhook_manager.go:
```diff
@@ -29,13 +30,14 @@ import (
 	configuration *atomic.Value
 	lister        admissionregistrationlisters.ValidatingWebhookConfigurationLister
-	hasSynced     func() bool
+	delayer       *synctrack.Delayer
```
Why did you make this a pointer?

lavalamp replied:
For consistency with the `configuration` member above, which also doesn't need to be a pointer...

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/synctrack/delayer_test.go:
```diff
@@ -0,0 +1,91 @@
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for i := 0; i < adds; i++ {
+			d.Add()
+		}
+	}()
```
Sure

lavalamp replied:
Done

deads2k commented:
having `updateConfiguration` handle the complete list of validating or mutating admission webhooks ensures 

1. order of updates to admission webhooks will always be honored. It is impossible to update-2 for webhook/B to take effect before update-1 for webhook/A to take effect.

But the only known failure mode is actually the lister.List call itself: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/configuration/validating_webhook_manager.go#L77-L84

With the new ability to ensure processing of the initial list has happened (your previous hasSync'd fix) and the synchronous processing done in the add/update/delete func, is there still value in having the `List` versus having special handling for delete and performing a running adding/remove, perhaps with a periodic sanity check for a diff with a lock?

deads2k commented:
> And waiting and then listing once is probably more efficient on startup than doing N inserts (not that it matters much for this code -- it's about the precedent for me)

Which precedent do we want?  Prior to your PR that tracks "have all event handlers handled", this was the state of the art.  After your PR, it is possible to build reliable tracking without the list and I think that combined with a periodic consistency checker, is likely how I would suggest people handle it.  Would you suggest using a list plus a delay instead?

lavalamp committed:
Fix N^2 startup for webhook configurations

Add a "lazy" type to track when an update is needed. It uses a nested
locking technique to avoid extra evaluation calls.

lavalamp commented:
Switched from the delay to lazy computation. (Incremental computation isn't efficient because I'd have to switch to a map from a list for intermediate storage and there would be a lot of copying.)

The tests got longer because they were broken previously (not actually testing the informer delivery notification) and I fixed them.

Where should the Lazy type go?

cici37 commented:
/triage accepted

apelisse commented on staging/src/k8s.io/client-go/tools/cache/synctrack/lazy.go:
```diff
@@ -0,0 +1,81 @@
+// current cached value is stale (due to a call to Notify), then Evaluate will
+// be called synchronously. If subsequent calls to Get happen, they will all
+// get the same return value (due to use of singleflight).
+func (z *Lazy[T]) Get() (T, error) {
```
I'm sure this code works but that seems excessively complicated. Why didn't you use a mutex rather than a combination of singleflight, atomic int64, and atomic value? Since eventually singleflight is still implemented with a mutex anyway?

lavalamp replied:
I've now pushed 2 commits; "if we use a lock instead" demonstrates the naive lock impl. The last commit shows the current solution but with singleflight rewritten out (fun exercise).

I also added a throughput test. The naive lock impl takes 7x as long as the impl in the final commit.

lavalamp replied:
oh, with the old singleflight impl, it took ~7.7s to run the tests 1000 times (sans throughput test). The current impl is a bit faster at ~7.1s.

logicalhan commented:
/remove-sig instrumentation

apelisse commented on staging/src/k8s.io/client-go/tools/cache/synctrack/lazy.go:
```diff
@@ -0,0 +1,83 @@
+func (z *Lazy[T]) Get() (T, error) {
+	e := z.cache.Load()
+	if e == nil {
+		// Since we don't force a constructor, nil is a possible value.
```
Good catche! (typo intended 😂 )

apelisse commented on staging/src/k8s.io/client-go/tools/cache/synctrack/lazy.go:
```diff
@@ -0,0 +1,81 @@
+	Evaluate func() (T, error)
+
+	counter int64
+	cache   atomic.Value
```
Yeah, that comment is already outdated, from an unpublished previous review ...

apelisse commented:
/lgtm
/approve
If you want to make the test external, I don't really care:
/hold

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/synctrack/lazy_test.go:
```diff
@@ -0,0 +1,248 @@
+limitations under the License.
+*/
+
+package synctrack
```
done!

lavalamp commented:
/hold cancel
